keps/420-partial-admission/README.md

```diff
@@ -0,0 +1,163 @@
+
+## Design Details
+
+### Workload API
+
+```go
+type PodSet struct {
+    // .......
+
+    // count is the number of pods for the spec.
+    // +kubebuilder:validation:Minimum=1
+    Count int32 `json:"count"`
+
+    // minimumCount is the minimum number of pods for the spec acceptable
+    // in case of partial admission.
+    //
+    // If not provided, partial admission for the current PodSet is not
+    // enabled.
+    // +optional
+    MinimumCount *int32 `json:"minimumCount,omitempty"`
+}
+
+```
+
+### Scheduler / Flavorassignment
+
+In case the workload proposed for the current scheduling cycle, does not fit, with or without preemption, in the current available quota and any of its PodSets allow partial admission, try to find to find a lower counts combination that fits the available quota.
+
+The search should be optimized (binary search) and preserve the proportion of pods lost across the variable count PodSets.
+
+Starting with the PodSets:
+
+```go
+[]podSet { {Count: 1} , {Count: 4, MinCount: 2}, {Count: 20, MinCount: 10}}
+```
+
+And only being able to admit 19 pods, it should end up with 
+
+
+```go
+[]podSet { {Count: 1} , {Count: 3}, {Count: 15}}
+```
+
+With both pods that accept partial admission losing 50% of their range.
+
+
+The accepted number of pods in each PodSet are recorded in `workload.Status.Admission.PodSetAssignments[*].ResourceUsage.Count`
+
+### Jobframework
+
+```diff
+type GenericJob interface {
+    // ...
+
+	
+-    // RunWithNodeAffinity will inject the node affinity extracting from workload to job and unsuspend the job.
++    // RunWithPodSetsInfo will inject the node affinity and podsSet counts extracted from workload to the job and unsuspend the job.
+-    RunWithNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RunWithPodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+-    // RestoreNodeAffinity will restore the original node affinity of job.
++    // RestorePodSetsInfo will restore the original node affinity of job.
+-    RestoreNodeAffinity(nodeSelectors []PodSetNodeSelector)
++    RestorePodSetsInfo(nodeSelectors []PodSetNodeSelector, podSetCounts []int32)
+
+    // ...
+}
+
+```
+
+### batch/Job controller
+
+Besides adapting `RunWithPodSetsInfo` and `RestorePodSetsInfo` it should also:
+
+- rework `PodSets()` to populate `MinimumCount` if the job is marked to support partial admission.
+  * jobs supporting partial admission should have a dedicated annotation. eg. `kueue.x-k8s.io/job-min-parallelism`, indicating the minimum `parallelism` acceptable by the job in case of partial admission.
+  * jobs which need the `completions` count kept in sync with `parallelism` should indicate this in a second annotation `kueue.x-k8s.io/sync-completions`
+- rework `EquivalentToWorkload` to account for potential differences in `PodSets` spec `Parallelism`.
+
+### kubeflow/MPIJob controller
```

Comment:
@ahg-g how would you think about this feature for JobSet?

Could there be multiple podsets that can scale? or can you also limit it to one template to be resizable?

Reply:
There are no restrictions currently, and I don't think we want to impose any in JobSet, there is no justification to do so.

Reply:
The question we have in the kueue side is that if we allow multiple podsets so scale, then we have an NP problem.
Hence the question of whether it is necessary to scale more than one podset, and if so, if there are any simplifications we can make.

So far, for Job, MPIJob and RayJob, there is only one podset that scales.

Reply:
Actually, RayJob has the concept of multiple worker specs. @trasc can you investigate or reach out to them to see how they handle resizing in their autoscaler?

Reply:
We should do it in #783